### PR TITLE
Add `Memory::data_and_store_mut`

### PR DESCRIPTION
**Pull request by AldaronLau**
Solves https://github.com/paritytech/wasmi/issues/447

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## CRITERION BENCHMARKS ## 

 |BENCHMARK|MASTER|PR|Diff|
 |---|---|---|---|
 | `compile_and_validate_v1`  | 5.4304 ms  | 5.4687 ms  | :white_circle:  +0.5215% |
| `execute_count_until_v1`  | 2.0224 ms  | 2.0243 ms  | :white_circle:  +0.1420% |
| `execute_factorial_iterative_v1`  | 917.67 ns  | 918.47 ns  | :white_circle:  +0.0417% |
| `execute_factorial_recursive_v1`  |  1.2501 µs   |  1.2479 µs   | :white_circle:  -0.1260% |
| `execute_fib_iterative_v1`  | 4.8018 ms  | 4.8009 ms  | :white_circle:  -0.0805% |
| `execute_fib_recursive_v1`  | 11.134 ms  | 11.124 ms  | :white_circle:  -0.0817% |
| `execute_global_bump_v1`  | 2.8145 ms  | 2.8134 ms  | :white_circle:  -0.0398% |
| `execute_host_calls_v1`  |  31.374 µs   |  32.080 µs   | :white_circle:  +0.4299% |
| `execute_memory_fill_v1`  | 4.0537 ms  | 4.0519 ms  | :white_circle:  -0.0357% |
| `execute_memory_sum_v1`  | 3.7995 ms  | 3.7974 ms  | :white_circle:  -0.0656% |
| `execute_memory_vec_add_v1`  | 8.0435 ms  | 8.0483 ms  | :white_circle:  +0.0771% |
| `execute_recursive_is_even_v1`  | 2.2321 ms  | 2.2317 ms  | :white_circle:  -0.0137% |
| `execute_recursive_ok_v1`  |  295.43 µs   |  295.33 µs   | :white_circle:  -0.0270% |
| `execute_recursive_scan_v1`  |  357.19 µs   |  357.35 µs   | :white_circle:  -0.0231% |
| `execute_recursive_trap_v1`  |  25.356 µs   |  25.375 µs   | :white_circle:  +0.1397% |
| `execute_regex_redux_v1`  | 1.4109 ms  | 1.3928 ms  | :white_circle:  -1.1911% |
| `execute_rev_complement_v1`  | 1.3896 ms  | 1.4095 ms  | :red_circle:  +1.5949% |
| `execute_tiny_keccak_v1`  | 1.2310 ms  | 1.2538 ms  | :red_circle:  +1.6914% |
| `execute_trunc_f2i_v1`  | 1.7962 ms  | 1.8039 ms  | :white_circle:  +0.2949% |
| `instantiate_v1`  |  75.089 µs   |  73.800 µs   | :white_circle:  -1.3501% |


 [Link to pipeline](https://gitlab.parity.io/parity/mirrors/wasmi/-/jobs/1852630)

File: wasmi_v1/src/store.rs [hunk 1]
```diff
@@ -299,6 +299,24 @@ impl<T> Store<T> {
             .unwrap_or_else(|| panic!("failed to resolve stored linear memory: {:?}", entity_index))
     }
 
+    /// Returns an exclusive reference to the associated entity of the linear memory and a shared
+    /// reference to the user provided state.
```
Comment (Robbepop) [inline wasmi_v1/src/store.rs:303]:
```suggestion
    /// Returns an exclusive reference to the associated entity of the linear memory and an exclusive
    /// reference to the user provided state.
```

**Review by Robbepop — COMMENTED** [review]
comments needs some small fix but besides that looks good to me! Thanks a lot for the PR!

**Comment by AldaronLau** [comment]
I pushed the comment fix; I'm wondering if the `Memory::data()` and `Memory::data_mut()` comments need to be changed as well;

> ``Returns a shared slice to the bytes underlying to the byte buffer.``

> ``Returns an exclusive slice to the bytes underlying to the byte buffer.``

If they were to match, it would say:

> ``Returns an shared slice to the bytes underlying to the [`Memory`].``

> ``Returns an exclusive slice to the bytes underlying to the [`Memory`].``

Also, I'm thinking grammatically, it would make sense to remove "to"

**Review by Robbepop — APPROVED** [review]

**Comment by Robbepop** [comment]
Want to fix the other comments before we merge this? I am also fine just merging.

**Comment by AldaronLau** [comment]
Sure, I can add the fixes.

**Comment by Robbepop** [comment]
Thanks a lot for the new API @AldaronLau !

**Pull request merged**